Not using _type to store class type (as _type is already deprecated in ES and will be removed soon)

ninodafonte opened the pull request:
Hey there 👋

**Context**

The _type field used to store the class mapping information for the read model objects is deprecated from 6.0 (more info [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html) and will be removed in 8.0.

Usually this field is used to store different types of objects in the same index. As this is not the recommended way from now on (ES recommends having different indexes for different types of objects), we should be able to find an alternative way of storing the mapping (our php object class reference).

**Proposal**

The issue started in #32 and this PR is a way of starting the conversation about what could be a good way to let users to decide if they want to use the _type field or not and, at the same time, being backwards compatible to not harm those already using the library.

Completely open to any other approach, suggestions, etc. 

Thanks!

othillo commented:
@ninodafonte Thank you so much for taking the time to look into this.

First of all, I don't mind breaking BC for this. We can tag a new major version that will not support types anymore. This will make your PR a lot simpler.

Second, I think it would be best to use 1 type per index as Elastic suggests. We sort of enforce that already by injecting the class as constructor parameter and checking the type in `save()`. Of course we can't prevent people creating another repository with the same index. 

I would indeed stop adding the type parameter when saving and not use it when finding. When deserializing, just use the class that's set in the repository class and assume it fails if someone does store a different class. Also try to add some tests to test behavior when you don't follow the 1 type per index best practice.

ninodafonte commented:
Hey Sander! 

After sorting out the pile of work that was waiting for me after vacation, I'm back ☝️
Thanks for the comments and the guidance, I did the following:

- Removed the flag to use or not the class type as we agree that we don't have to keep it BC.
- Adapted the tests and the save method to reflect the fact that different types in the same repo are not allowed.
- Fixed a bunch of php_cs_fixes and test naming based on the definitions you have in the checks (I assume you changed them along the way becase most of the test code was using a different set of rules, test naming for example).

So, let me know how does it look like now and we'll keep going, 

thanks for your time!

othillo commented on src/ElasticSearchRepository.php:
```diff
@@ -120,13 +122,15 @@ public function findAll(): array
     {
         try {
-            $this->client->delete([
+            $params = [
```
Personally, I try to avoid creating one-off local variables.

ninodafonte replied:
Sure thing, I'll change it. I've got used to do it that way sometimes to help me debug

othillo commented on test/ElasticSearchRepositoryFactoryTest.php:
```diff
@@ -22,7 +22,7 @@ class ElasticSearchRepositoryFactoryTest extends TestCase
      * @test
      */
-    public function it_creates_an_elastic_search_repository()
+    public function itCreatesAnElasticSearchRepository(): void
```
hmm, this is possible enforced by a newer version of php-cs-fixer. In `broadway/broadway` we prevented this by https://github.com/broadway/broadway/blob/master/.php_cs#L14-L16.

I like to keep the test methods names in snake case.

ninodafonte replied:
Yep, I prefer snake case as well, just got surprised but the phpcs rules, I'll get them back

othillo commented on src/ElasticSearchRepository.php:
```diff
@@ -44,28 +46,29 @@ public function __construct(
     public function save(Identifiable $data): void
     {
-        Assertion::isInstanceOf($data, $this->class);
+        Assertion::true(
```
Does the assertion library not support providing error messages for the `isInstanceOf` test?

ninodafonte replied:
It does indeed. I was tricked by the PHP Inspections of the IDE recommending me this option to avoid some edge case with the string type of the class. Never mind, getting it back :)

ninodafonte replied:
More info coming from the phpstan analyzer: 
 67     Parameter #2 $className of static method Assert\Assertion::isInstanceOf() expects class-string<object>, string given.

othillo reviewed (commented):
Nice work, exactly what we discussed. Some minor comments.

ninodafonte commented on src/ElasticSearchRepository.php:
```diff
@@ -57,24 +58,31 @@ public function __construct(
+            $data,
+            $this->class_type,
             "Data object should be of type {$this->class_type}, as declared on the repository definition."
         );
```
I've found out after some reading that this is what leaves phpstan confident that you know what you are doing. It adds an extra check, what do you think about it?

More info: https://github.com/phpstan/phpstan/issues/2808

othillo commented on src/ElasticSearchRepository.php:
```diff
@@ -44,34 +47,42 @@ public function __construct(
     public function save(Identifiable $data): void
     {
-        Assertion::isInstanceOf($data, $this->class);
+        if (!class_exists($this->class_type)) {
```
If we want to check if this class exists, we should do it in the constructor.

ninodafonte replied:
:) My thoughts exactly, but then phpstan is not aware of the check and raise the exception again

othillo commented on src/ElasticSearchRepository.php:
```diff
@@ -44,34 +47,42 @@ public function __construct(
+            );
+        }
+
+        Assertion::isInstanceOf(
```
The more I think of this the more I favor a custom exception for this. Maybe later.

ninodafonte commented on src/ElasticSearchRepository.php:
```diff
@@ -44,34 +45,35 @@ public function __construct(
-        Assertion::isInstanceOf($data, $this->class);
+        if (!$data instanceof $this->class_type) {
+            throw new MultiTypeIndexNotAllowedException(
+                "Data object should be of type {$this->class_type}, as declared on the repository definition."
+            );
+        }
 
         $serializedReadModel = $this->serializer->serialize($data);
-
-        $params = [
+        $this->client->index([
```
So, giving it a step back and using a custom exception, how does it look like now? phpstan is happy with it

ninodafonte reviewed (commented):
another option

othillo commented:
Thank you, @ninodafonte !
